src/common/components/markdownViewer/markdownViewerComponent.tsx

```diff
@@ -1,33 +1,110 @@
-  let childrenComponent: React.ReactNode = null;
+class MarkDownViewer extends React.Component<MarkDownViewerComponentProps, {}> {
+  constructor(props) {
+    super(props);
+  }
 
-  if (content) {
-    childrenComponent = compile(content).tree;
+  private CreateMdrInstance = () => {
+    const setupParams: MdrSetup = {
+      routerLocation: this.props.location ? this.props.location.pathname : '',
+      options: this.props.options,
+      codeStyle: this.props.codeStyle,
+    };
+    return CreateMarkdownRender(setupParams);
   }
 
-  return childrenComponent;
-};
+  private mdr: Mdr = this.CreateMdrInstance();
+  private scrollableContainerRef: HTMLElement = null;
+  private allowNotifyScroll: boolean = true;
+
+  private setScrollableContainerRef = (input) => { this.scrollableContainerRef = input; };
+
+  private markdownToHTML = () => {
+    return {
+      __html: this.mdr.render(this.props.content || ''),
+    };
+  }
+
+  private doScrollToSourceLine = (targetSourceLine) => {
+    const renderedElements = ReactDOM.findDOMNode(this).getElementsByClassName(SOURCE_LINE_CLASSNAME);
+    const scrollOffset = getPixelOffsetForSourceLine(renderedElements, targetSourceLine);
+    const componentPosition = this.scrollableContainerRef.getBoundingClientRect().top;
+    this.allowNotifyScroll = false;
+    this.scrollableContainerRef.scrollTop += scrollOffset - ((componentPosition > 0) ? componentPosition : 0);
+  }
+
+  private notifySourceLine = throttle(() => {
+    const componentPosition = this.scrollableContainerRef.getBoundingClientRect().top;
+    const renderedElements = ReactDOM.findDOMNode(this).getElementsByClassName(SOURCE_LINE_CLASSNAME);
+    const lineNum = getSourceLineForPixelOffset(renderedElements, componentPosition > 0 ? componentPosition : 0);
+    this.props.onScrollSourceLine(lineNum);
+  }, 25);
 
-export const MarkDownViewerComponent: React.StatelessComponent<MarkDownViewerComponentProps> = ({ content }) => {
-  return (
-    <div>
-      {getMarkDownChildren(content)}
-    </div>
-  );
+  private handleScroll = (event) => {
+    if (this.allowNotifyScroll) {
+      if (this.props.onScrollSourceLine) {
+        this.notifySourceLine();
+      }
+    } else {
+      this.allowNotifyScroll = true;
+    }
+  }
+
+  public shouldComponentUpdate(nextProps, nextState) {
+    if (nextProps.scrollSourceLine !== undefined && nextProps.scrollSourceLine !== this.props.scrollSourceLine) {
+      this.doScrollToSourceLine(nextProps.scrollSourceLine);
+      return false;
+    }
+    return true;
+  }
+
+  public componentWillUpdate(nextProps, nextState) {
+    // New render instance upon render options update.
+    if (nextProps.options !== this.props.options) {
+      this.mdr = this.CreateMdrInstance();
+    }
+  }
+
+  public render() {
+    // Object destructuring to retrieve className with default value.
+    const {className = ''} = this.props;
+    // WARNING: This conversion from plain HTML to JSX with
```

Comment:
What about this warning?

Reply:
Yes it is. You can check it [here](https://github.com/markdown-it/markdown-it/blob/master/docs/security.md). I disabled HTML tags intentionally.